internal/app/orchestrator/orchestrator.go

```diff
@@ -0,0 +1,113 @@
+// Package orchestrator is responsible for instrumenting inbound xDS client
+// requests to the correct aggregated key, forwarding a representative request
+// to the upstream origin server, and managing the lifecycle of downstream and
+// upstream connections and associates streams. It implements
+// go-control-plane's Cache interface in order to receive xDS-based requests,
+// send responses, and handle gRPC streams.
+package orchestrator
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/envoyproxy/xds-relay/internal/app/cache"
+	"github.com/envoyproxy/xds-relay/internal/app/mapper"
+	"github.com/envoyproxy/xds-relay/internal/app/upstream"
+	"github.com/envoyproxy/xds-relay/internal/pkg/log"
+	yamlproto "github.com/envoyproxy/xds-relay/internal/pkg/util"
+	aggregationv1 "github.com/envoyproxy/xds-relay/pkg/api/aggregation/v1"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	gcp "github.com/envoyproxy/go-control-plane/pkg/cache/v2"
+)
+
+const (
+	component = "orchestrator"
+
+	// TODO (https://github.com/envoyproxy/xds-relay/issues/41) load from configured defaults.
+	aggregationRules  = ""
+	cacheMaxEntries   = 1000
+	cacheTTL          = 60 * time.Minute
+	upstreamClientURL = "localhost:8080"
+)
+
+// Orchestrator has the following responsibilities:
+//
+// 1. Aggregates similar requests abiding by the aggregated keyer
+//    configurations.
+// 2. Maintains long lived streams with the upstream origin server for each
+//    representative discovery request.
+// 3. Maintains long lived streams with each downstream xDS client using
+//    go-control-plane's CreateWatch function.
+// 4. When new responses are available upstream, orchestrator relays and fans
+//    out the response back on the streams associated with the xDS clients.
+// 5. Updates the xds-relay cache with the latest state of the world.
+type Orchestrator interface {
+	gcp.Cache
+}
+
+type orchestrator struct {
+	mapper         mapper.Mapper
+	cache          cache.Cache
+	upstreamClient upstream.Client
+
+	logger log.Logger
+}
+
+// New instantiates the mapper, cache, upstream client components necessary for
+// the orchestrator to operate and returns an instance of the instantiated
+// orchestrator.
+func New(ctx context.Context, l log.Logger) Orchestrator {
+	orchestrator := &orchestrator{
+		logger: l.Named(component),
+	}
+
+	cache, err := cache.NewCache(cacheMaxEntries, orchestrator.onCacheEvicted, cacheTTL)
+	if err != nil {
+		orchestrator.logger.With("error", err).Panic(ctx, "failed to initialize cache")
+	}
+
+	upstreamClient, err := upstream.NewClient(ctx, upstreamClientURL)
+	if err != nil {
+		orchestrator.logger.With("error", err).Panic(ctx, "failed to initialize upstream client")
+	}
+
+	var config aggregationv1.KeyerConfiguration
+	err = yamlproto.FromYAMLToKeyerConfiguration(aggregationRules, &config)
```

Comment:
Sounds fair, I'll make the change. What are your thoughts on who initializes the mapper and the xDS client?

Reply:
i think they should happen outside of orchestrator too. We could call it an initializer, which parses the flags, creates the singletons, etc

Reply:
The orchestrator is complex enough, the intention is to remove anything that is not strictly orchestrator related.

Reply:
Cache takes a onCacheEvicted function callback. I think this makes it more tricky to initialize outside of the orchestrator, because the orchestrator "owns" the eviction workflow.